### PR TITLE
Added a new fast implementation of the EPL profile

### PR DESCRIPTION
**Pull request by ewoudwempe**
This is the same profile as the already existing EPL profile, but my own implementation of it with Numba. Numba code is never pretty, but it is quite a bit faster than the existing PEMD models. Timings:
https://gist.github.com/ewoudwempe/10d47a7433afdb4d750dfad29e8dcd95
Comments? I suppose I'll have to make some tests too...
I could cache the parameter conversions, but it is a very small part of the computation times, so I don't think it's worth it.

### CODE REVIEW

File: lenstronomy/LensModel/Profiles/epl_numba.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+__all__ = ['EPL_numba']
+
+
+class EPL_numba(LensProfileBase):
+    """
+    class for power law ellipse mass density profile. Follows the formulas from Tessore+ (2015).
+
+    The Einstein ring parameter converts to the definition used by GRAVLENS as follow:
+    (theta_E / theta_E_gravlens) = sqrt[ (1+q^2) / (2 q) ]
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_numba.py:17]: I try to transition all formulas to latex and rendering in the documentation, primarily by requiring PRs and new code to do so. I'll do it right away for the EPL class and then you can adopt it here. Should be done soon on my side

File: lenstronomy/LensModel/Profiles/epl_numba.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        f_xx = kappa + gamma_shear.real
+        f_yy = kappa - gamma_shear.real
+        f_xy = gamma_shear.imag
+        # Fix the nans if x=y=0 is filled in
+
+        return nan_to_zero(f_xx), nan_to_zero(f_yy), nan_to_zero(f_xy)
+
+@jit()
+def param_transform(x, y, theta_E, gamma, e1, e2, center_x=0., center_y=0.):
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_numba.py:101]: can you add a description of the parameters?

File: lenstronomy/LensModel/Profiles/epl_numba.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    t = gamma-1
+    phi_G, q = param_util.ellipticity2phi_q(e1, e2)
+    x_shift = x - center_x
+    y_shift = y - center_y
+    ang = phi_G
+    z = np.exp(-1j*phi_G) * (x_shift + y_shift*1j)
+    return z, theta_E*np.sqrt(q), t, q, ang
+
+
+@nb.generated_jit(nopython=True, cache=True)
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_numba.py:112]: I am a bit worried about the fixed settings nopython=True and cache=True. Perhaps you can move this routine to the numba_util.py file and demand the same environment configurations as for the other @jit decorators.

File: lenstronomy/Util/param_util.py [hunk 4]
```diff
@@ -69,6 +70,7 @@ def shear_cartesian2polar(gamma1, gamma2):
 
 
 @export
+@jit
```
Comment (sibirrer) [inline lenstronomy/Util/param_util.py]: I see that you need those. Given that some other people manually out-commented previous decorators as a hack-fix, I'd like to first approve the PR with the fall-back and/or configurable numba settings

**Review by sibirrer — CHANGES_REQUESTED** [review]
Overall looks great, and thanks also for the comprehensible testing! It really makes a difference. The tests failed (due to numba settings I guess). Perhaps you can move the @jit decorator on the ellipticity2phi_q transforms to your other PR dedicated to numba settings.

What is missing is test routines. You could add tests cross-comparing with the hyperbolic function implementation in 'EPL' and the numerical derivative tests: https://github.com/sibirrer/lenstronomy/blob/master/test/test_LensModel/test_numeric_lens_differentials.py
Thanks a lot and let me know if you have questions!

File: test/test_LensModel/test_numeric_lens_differentials.py [hunk 5]
```diff
@@ -266,6 +266,11 @@ def test_EPL(self):
         lens_model = ['EPL']
         self.assert_differentials(lens_model, kwargs)
 
+    def test_EPL_numba(self):
+        kwargs = {'theta_E': 2., 'e1': 0.1, 'e2': 0., 'gamma': 2.13}
+        lens_model = ['EPL']
```
Comment (sibirrer) [inline test/test_LensModel/test_numeric_lens_differentials.py:271]: I guess here you wanted the EPL_NUMBA model to be tested

File: lenstronomy/Util/param_util.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import numpy as np
 
+from lenstronomy.Util.numba_util import jit
 from lenstronomy.Util.package_util import exporter
+from lenstronomy.Util.numba_util import jit
```
Comment (sibirrer) [inline lenstronomy/Util/param_util.py:5]: I guess this is double imported

File: lenstronomy/Util/numba_util.py [hunk 7]
```diff
@@ -38,20 +37,64 @@
...
         def wrapper(func):
-            return numba.jit(func, nopython=nopython, cache=cache, parallel=parallel, fastmath=fastmath,
-                             error_model=error_model)
+            return numba.jit(func, nopython=nopython, cache=cache, parallel=parallel,fastmath=fastmath,error_model=error_model)
+    else:
+        def wrapper(func):
+            return func
+    return wrapper
+
+def generated_jit(nopython=nopython, cache=cache, parallel=parallel,fastmath=fastmath,error_model=error_model):
```
Comment (sibirrer) [inline lenstronomy/Util/numba_util.py:58]: can you add a short description of this function and the definition arguments?

File: lenstronomy/Util/numba_util.py [hunk 8]
```diff
@@ -38,20 +37,64 @@
...
+    """Implements an equivalent to np.nan_to_num (with copy=False!) array or scalar in Numba.
+    The generated_jit part is necessary because of the need to support both arrays and scalars for all input functions.
+    """
+    if (isinstance(x, numba.types.Array) or isinstance(x, np.ndarray)) and x.ndim > 0:
+        return nan_to_num_arr if numba_enabled else nan_to_num_arr(x, posinf, neginf, nan)
+    else:
+        return nan_to_num_single if numba_enabled else nan_to_num_single(x, posinf, neginf, nan)
+
+@jit()
+def nan_to_num_arr(x, posinf=1e10, neginf=-1e10, nan=0.):
```
Comment (sibirrer) [inline lenstronomy/Util/numba_util.py:79]: I guess it's clear what this routine does but for better visibility in the documentation, can you add docstrings?

File: lenstronomy/Util/numba_util.py [hunk 9]
```diff
@@ -38,20 +37,64 @@
...
+            x[i] = nan
+        if np.isinf(x[i]):
+            if x[i] > 0:
+                x[i]=posinf
+            else:
+                x[i]=neginf
+    return x
+
+@jit()
+def nan_to_num_single(x, posinf=1e10, neginf=-1e10, nan=0.):
```
Comment (sibirrer) [inline lenstronomy/Util/numba_util.py:91]: I guess it's clear what this routine does but for better visibility in the documentation, can you add docstrings?

File: test/test_LensModel/test_Profiles/test_epl_numba.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        phi_E = 1.
+        gamma = 2.
+        q = 1.
+        phi_G = 1.
+        e1, e2 = param_util.phi_q2_ellipticity(phi_G, q)
+
+        #x = 0.
+        #y = 0.
+        #f_x, f_y = self.EPL_numba.derivatives(x, y, phi_E, e1, e2, gamma)
+        #npt.assert_almost_equal(f_x, 0.)
```
Comment (sibirrer) [inline test/test_LensModel/test_Profiles/test_epl_numba.py:100]: is there a reason why this fails? (frankly, I have not tested that for either the SPEMD nor EPL model) It's fine if it's not tested as long as (I see that) the magnification is basically zero

**Review by sibirrer — CHANGES_REQUESTED** [review]
This looks great! I have only very minor comments. I guess the other PR becomes obsolete then. I can try to turn off numba in the test coverage on my side indeed and this is not part of this PR. Thanks a lot!

File: test/test_LensModel/test_Profiles/test_epl_numba.py (hunk 10 above)
Comment (ewoudwempe) [reply test/test_LensModel/test_Profiles/test_epl_numba.py:100]: Ah, it actually runs fine, I forgot to add that one back

**Comment by ewoudwempe** [comment]
Everything passes :)
I think I addressed all your comments, but let me know otherwise!

**Comment by sibirrer** [comment]
> Everything passes :)
> I think I addressed all your comments, but let me know otherwise!

Excellent! If you add the documentation strings for the definitions in the other PR (and probably sync it with this one not to create merge conflicts when merging one and then the other), we are done with it.

**Review by sibirrer — APPROVED** [review]

**Comment by ewoudwempe** [comment]
Something went wrong here... let me try to fix...

**Pull request merged**